k8s/pkg/registryserver/registry_cache.go

```diff
@@ -134,27 +134,43 @@ func (rc *registryCacheImpl) GetEndpointsByNsm(nsmName string) []*v1.NetworkServ
 	return rc.networkServiceEndpointCache.GetByNetworkServiceManager(nsmName)
 }
 
+const maxAllowedAttempts = 10
+
 func (rc *registryCacheImpl) CreateOrUpdateNetworkServiceManager(nsm *v1.NetworkServiceManager) (*v1.NetworkServiceManager, error) {
 	existingNsm := rc.networkServiceManagerCache.Get(nsm.GetName())
 
-	if existingNsm != nil {
-		nsm.ObjectMeta = existingNsm.ObjectMeta
-		logrus.Infof("NSM with name %v already exist in cache, updating: %v", nsm.GetName(), nsm)
-		return rc.updateNetworkServiceManager(nsm)
-	}
-
-	logrus.Infof("Creating NSM: %v", nsm)
-	createNsm, err := rc.addNetworkServiceManager(nsm)
-	if err != nil || apierrors.IsAlreadyExists(err) {
-		existingNsm, err = rc.clientset.NetworkservicemeshV1().NetworkServiceManagers(rc.nsmNamespace).Get(nsm.Name, metav1.GetOptions{})
-		if err != nil {
-			return existingNsm, err
+	attempt := 0
+	for attempt < maxAllowedAttempts {
+		logrus.Infof("CreateOrUpdateNSM attempt %d: ", attempt)
+
+		if existingNsm == nil {
+			logrus.Infof("Creating NSM: %v", nsm)
+			newNsm, err := rc.addNetworkServiceManager(nsm)
+			if err == nil || !apierrors.IsAlreadyExists(err) {
+				return newNsm, err
+			}
+
+			logrus.Infof("NSM with name %v already exist", nsm.GetName())
+			existingNsm, err = rc.GetNetworkServiceManager(nsm.GetName())
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			logrus.Infof("Updating existing NSM: %v with %v", existingNsm, nsm)
+			updNsm := nsm.DeepCopy()
+			updNsm.ObjectMeta = existingNsm.ObjectMeta
+			updNsm, err := rc.updateNetworkServiceManager(updNsm)
+			if err == nil || !apierrors.IsConflict(err) {
+				return updNsm, err
+			}
+
+			logrus.Infof("There is no NSM with name %v", nsm.GetName())
+			existingNsm = nil
```

Comment:
whole this `if .. else ...` looks very strange. I am sure it can be rewritten in a clearer fashion.
So for example here you are in the branche where there is `existingNsm != nil` and then you find that there is no NSMgr with that name. Is that even possible?
Please rethink the logic here.

Reply:
Yes, that exactly what happened sometimes in tests. Previous test sent DELETE to k8s and it was handled by k8s with some delay. So we have NSM in cache, but k8s-registry already empty, so Update will fail. 

About rewritting, actually I found this way of implement that logic the most readable. But I'll think about others ways more carefully